python/lsst/db/dbPool.py

```diff
@@ -0,0 +1,100 @@
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the LSST License Statement and
+# the GNU General Public License along with this program.  If not,
+# see <http://www.lsstcorp.org/LegalNotices/>.
+
+"""
+This module implement connection pool - a pool of "db" objects.
+
+@author  Jacek Becla, SLAC
+
+"""
+
+# standard library imports
+import logging as log
+from time import time
+
+# local imports
+from lsst.db.db import Db
+from lsst.db.exception import produceExceptionClass
+
+####################################################################################
+
+DbPoolException = produceExceptionClass('DbException', [
+        (1600, "ENTRY_NOT_FOUND", "Requested Db entry not found.")])
+
+####################################################################################
+class DbPool(object):
+    """
+    @brief A pool of Db objects.
+
+    This class implements a pool of Db objects. Optionally, it can maintain live
+    connection and automatically recover from time outs.
+    """
+
+    def __init__(self):
+        """
+        Create a DbPool instance.
+        """
+        self._pool = {}
+
+    ##### Connection-related functions #############################################
+    def addConn(self, cName, dbConn, checkFreq=600):
+        """
+        Add Db Connection Object to the pool.
+
+        @param cName      Name of this connection
+        @param dbConn     Db Connection Object
+        @param checkFreq  Frequency of rechecking if connection is alive in seconds,
+                          -1 for "don't check it at all"
+        """
+        # Remember timestamp when connection was last checked.
+        lastChecked = 0
+
+        # the pool of named Db objects along with checkFreq and lastCheck time
+        self._pool[cName] = (dbConn, checkFreq, lastChecked)
+
+    def delConn(self, cName):
+        """
+        Remove Db Connection Object corresponding to a given name from the pool.
+        """
+        if cName in self._pool:
+            del self._pool[cName]
+
+    def getConn(self, cName):
+        """
+        Return Db Connection Object corresponding to a given name, optionally
+        ensure connection did not time out and reconnect if needed.
+
+        If will raise ENTRY_NOT_FOUND exception if the entry is not found for a
+        given name.
+        """
+        if cName not in self._pool:
+            raise DbPoolException(DbPoolException.ENTRY_NOT_FOUND, cName)
+
+        (db, checkFreq, lastChecked) = self._pool[cName]
+        if checkFreq != -1:
+            if time() - lastChecked > checkFreq:
+                log.debug("Checking connection before executing query.")
```

Comment:
This will log to root logger, and there is little additional information to identify which connection you are processing. Consider logging it to non-root logger and adding connection name. Same for debug() below.